File: articles/classification-taxonomy-ordering-of-life/r/sequence_distance.R
# Sequence-distance workflow in R.

seq_path <- file.path("data", "aligned_sequences.csv")
if (!file.exists(seq_path)) {
  seq_path <- file.path("..", "data", "aligned_sequences.csv")
}

seq_df <- read.csv(seq_path)

p_distance <- function(a, b) {
  a_chars <- strsplit(a, "")[[1]]
  b_chars <- strsplit(b, "")[[1]]
  if (length(a_chars) != length(b_chars)) stop("Sequences must be aligned.")
  sum(a_chars != b_chars) / length(a_chars)
}

jukes_cantor <- function(p) {
  if (p >= 0.75) return(NA_real_)
  -0.75 * log(1 - (4 / 3) * p)
}

taxa <- seq_df$taxon
p_mat <- matrix(0, nrow = length(taxa), ncol = length(taxa), dimnames = list(taxa, taxa))
jc_mat <- p_mat

for (i in seq_along(taxa)) {
  for (j in seq_along(taxa)) {
    p <- p_distance(seq_df$sequence[i], seq_df$sequence[j])
    p_mat[i, j] <- p
    jc_mat[i, j] <- jukes_cantor(p)
  }
}

print(round(p_mat, 4))
print(round(jc_mat, 4))

File: articles/classification-taxonomy-ordering-of-life/python/sequence_distance.py
"""
Sequence-distance workflow.

Run:
    python python/sequence_distance.py
"""

from __future__ import annotations

from itertools import product
from pathlib import Path

import pandas as pd

from taxonomy_core import jukes_cantor_distance, p_distance


ARTICLE_DIR = Path(__file__).resolve().parents[1]
SEQ_PATH = ARTICLE_DIR / "data" / "aligned_sequences.csv"


def main() -> None:
    seq_df = pd.read_csv(SEQ_PATH)
    seqs = dict(zip(seq_df["taxon"], seq_df["sequence"]))
    taxa = list(seqs.keys())

    p_mat = pd.DataFrame(index=taxa, columns=taxa, dtype=float)
    jc_mat = pd.DataFrame(index=taxa, columns=taxa, dtype=float)

    for t1, t2 in product(taxa, taxa):
        p = p_distance(seqs[t1], seqs[t2])
        p_mat.loc[t1, t2] = p
        jc_mat.loc[t1, t2] = jukes_cantor_distance(p)

    print("Raw p-distance matrix:")
    print(p_mat.round(4).to_string())

    print("\nJukes-Cantor distance matrix:")
    print(jc_mat.round(4).to_string())


if __name__ == "__main__":
    main()
